Installation issue: py-torchvision: PyTorch has CUDA Version=11.7 and torchvision has CUDA Version=11.8

On a system with P100 GPUs, with the following Spack environment (where I explicitely set `cuda_arch=60` for some packages):

```yaml
spack:
  config:
    install_tree: /my-spack/spack
    build_stage: /tmp/spack-stage
  repos:
  - /home/tbouvier/spack-lille/mochi-spack-packages
  packages:
    all:
      providers:
        mpi: [openmpi]
      compiler:
      - gcc@10.4.0
    magma:
      variants: [+cuda cuda_arch=60]
    nccl:
      variants: [+cuda cuda_arch=60]
    nccl-tests:
      variants: [+cuda cuda_arch=60]
    py-horovod:
      variants: [+cuda cuda_arch=60 tensor_ops=nccl]
    py-torch:
      variants: [+cuda cuda_arch=60 +nccl +cudnn +magma +distributed]
  specs:
  # Precedence is important. These packages will be picked up in the mochi repo.
  - mochi-thallium ^argobots ^mercury +boostsys +ofi +debug ^libfabric +cuda +debug fabrics=tcp,rxm,verbs
  - cuda
  - cudnn@8.4
  - mpi
  - nccl
  - nccl-tests
  - python@3.10
  - nvtx +python
  - py-pip
  - py-pybind11
  - py-ruamel-yaml
  - py-filelock
  - py-wandb
  - py-continuum
  - py-timm
  - py-nvidia-dali@1.27.0-cuda110
  - py-torch@2.0.1 ~gloo
  - py-torchsummary
  - py-bokeh
  - py-matplotlib
  - py-horovod@0.28 controllers=mpi frameworks=pytorch
  - nano
  concretizer:
    unify: true
  view: /my-spack/view/chifflot-p100
```

From the error I get at runtime, it seems that there is a discrepancy between CUDA versions used to compile PyTorch and torchvision. I don't understand how this can happen, as I'm using `unify: true` in my environment. Could this be caused by [this](https://github.com/spack/spack/blob/da10487219ccd3b738b910000fe307de3fc4fdbd/var/spack/repos/builtin/packages/py-torchvision/package.py#L151) not picking up the correct `cuda_arch` for torchvision?

<pre>
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "/root/distributed-continual-learning/main.py", line 21, in <module>
[1,0]<stderr>:    import backbone
[1,0]<stderr>:  File "/root/distributed-continual-learning/backbone/__init__.py", line 2, in <module>
[1,0]<stderr>:    from .ghostnet import *
[1,0]<stderr>:  File "/root/distributed-continual-learning/backbone/ghostnet.py", line 4, in <module>
[1,0]<stderr>:    import timm
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/timm/__init__.py", line 2, in <module>
[1,0]<stderr>:    from .layers import is_scriptable, is_exportable, set_scriptable, set_exportable
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/timm/layers/__init__.py", line 11, in <module>
[1,0]<stderr>:    from .conv_bn_act import ConvNormAct, ConvNormActAa, ConvBnAct
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/timm/layers/conv_bn_act.py", line 9, in <module>
[1,0]<stderr>:    from .create_norm_act import get_norm_act_layer
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/timm/layers/create_norm_act.py", line 14, in <module>
[1,0]<stderr>:    from .norm_act import BatchNormAct2d, GroupNormAct, LayerNormAct, LayerNormAct2d
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/timm/layers/norm_act.py", line 20, in <module>
[1,0]<stderr>:    from torchvision.ops.misc import FrozenBatchNorm2d
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/__init__.py", line 6, in <module>
[1,0]<stderr>:    from torchvision import datasets, io, models, ops, transforms, utils
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/datasets/__init__.py", line 1, in <module>
[1,0]<stderr>:    from ._optical_flow import FlyingChairs, FlyingThings3D, HD1K, KittiFlow, Sintel
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/datasets/_optical_flow.py", line 12, in <module>
[1,0]<stderr>:    from ..io.image import _read_png_16
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/io/__init__.py", line 8, in <module>
[1,0]<stderr>:    from ._load_gpu_decoder import _HAS_GPU_VIDEO_DECODER
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/io/_load_gpu_decoder.py", line 1, in <module>
[1,0]<stderr>:    from ..extension import _load_library
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/extension.py", line 107, in <module>
[1,0]<stderr>:    _check_cuda_version()
[1,0]<stderr>:  File "/my-spack/view/chifflot-p100/lib/python3.10/site-packages/torchvision/extension.py", line 80, in _check_cuda_version
[1,0]<stderr>:    raise RuntimeError(
[1,0]<stderr>:RuntimeError: Detected that PyTorch and torchvision were compiled with different CUDA versions. PyTorch has CUDA Version=11.7 and torchvision has CUDA Version=11.8. Please reinstall the torchvision that matches your PyTorch install.
</pre>

Some information on my system:

* **Spack:** 0.21.0.dev0
* **Python:** 3.10.13
* **Platform:** linux-debian11-skylake_avx512
* **Concretizer:** clingo

Removing `cmake_prefix_paths` didn't help.

I tried with PyTorch version 2.1: it doesn't even compile anymore, but it catches my error early:

```
  CMake Error at cmake/public/cuda.cmake:65 (message):
    Found two conflicting CUDA installs:                                                                                                          
    V11.8.89 in
    '/grid5000/spack/v1/opt/spack/linux-debian11-x86_64_v2/gcc-10.4.0/cuda-11.7.1-mniz3r7mpcnkeb2ekh7kcscflfid42om/include'
    and
                                                                                                                                           
    V11.8.89 in
    '/my-spack/spack/chifflot-p100/linux-debian11-skylake_avx512/gcc-10.4.0/cuda-11.8.0-my6q3bskrb5y6cjgvq554cqok3vrwxyp/include'
  Call Stack (most recent call first):
    cmake/Dependencies.cmake:44 (include)
    CMakeLists.txt:722 (include)

```

[spack-build-out.txt](https://github.com/spack/spack/files/13244343/spack-build-out.txt)
[spack-build-env-mods.txt](https://github.com/spack/spack/files/13244615/spack-build-env-mods.txt)

For some reason CMake is picking up the CUDA install in `'/grid5000/spack/v1/opt/spack/linux-debian11-x86_64_v2/gcc-10.4.0/`. 

I'd like to try setting `CUDA_HOME`. How should I do that?

--- a/var/spack/repos/builtin/packages/py-torch/package.py
+++ b/var/spack/repos/builtin/packages/py-torch/package.py
@@ -498,9 +498,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
 
         enable_or_disable("cuda")
         if "+cuda" in self.spec:
-            # cmake/public/cuda.cmake
-            # cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake
-            env.unset("CUDA_ROOT")
+            env.set("CUDA_HOME", self.spec["cuda"].prefix)  # Linux/macOS
+            env.set("CUDA_PATH", self.spec["cuda"].prefix)  # Windows
             torch_cuda_arch = ";".join(
                 "{0:.1f}".format(float(i) / 10.0) for i in self.spec.variants["cuda_arch"].value
             )

